classes/lib.php

```diff
@@ -0,0 +1,115 @@
+<?php
+// This file is part of Moodle - https://moodle.org/
+//
+// Moodle is free software: you can redistribute it and/or modify
+// it under the terms of the GNU General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+//
+// Moodle is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU General Public License for more details.
+//
+// You should have received a copy of the GNU General Public License
+// along with Moodle.  If not, see <http://www.gnu.org/licenses/>.
+
+namespace tool_dataflows;
+
+/**
+ * Support functions for dataflows.
+ *
+ * @package   tool_dataflows
+ * @author    Jason den Dulk <jasondendulk@catalyst-au.net>
+ * @copyright 2022, Catalyst IT
+ * @license   http://www.gnu.org/copyleft/gpl.html GNU GPL v3 or later
+ */
+class lib {
+    /**
+     * Get the scheme for a path string. will default to 'file' if none present.
+     *
+     * @param string $path
+     * @return string
+     */
+    public static function path_get_scheme(string $path): string {
+        $splitpath = explode('://', $path, 2);
+        if (count($splitpath) === 2) {
+            return $splitpath[0];
+        } else {
+            return 'file';
+        }
+    }
+
+    /**
+     * Does the path string include a scheme?
+     *
+     * @param string $path
+     * @return bool
+     */
+    public static function path_has_scheme(string $path): bool {
+        $splitpath = explode('://', $path, 2);
+        return (count($splitpath) === 2);
+    }
+
+    /**
+     * Is the path a relative one?
+     *
+     * @param string $path
+     * @return bool
+     */
+    public static function path_is_relative(string $path): bool {
+        if (self::path_has_scheme($path)) {
+            return false;
+        }
+
+        // Unix absolute path.
+        $first = substr($path, 0, 1);
+        if ($first == '/' || $first == '\\') {
+            return false;
+        }
+
+        // Windows absolute path.
+        if (preg_match('/^[a-zA-Z]:\\\/', $path) === 1) {
+            return false;
+        }
+
+        return true;
+    }
+
+    /**
+     * Makes a full path from a relative one using th given base dir.
+     *
+     * @param string $path
+     * @param string $scratchdir
+     * @return string
+     */
+    public static function path_make_absolute(string $path, string $scratchdir): string {
+        if (self::path_is_relative($path)) {
+            return $scratchdir . '/' . $path;
+        } else {
+            return $path;
+        }
```

Comment:
Suggestion is to remove this `else` nesting as it is redundant
```suggestion

        return $path;
```